### PR TITLE
[Custom Extension] Fix custom double_grad backward=None

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
Bug fixes

### PR changes
Others

### Describe
Pre-PR fixes custom op bug of backward: [[Custom operator] Fix custom operator backward=None bug #48656](https://github.com/PaddlePaddle/Paddle/pull/48656)
This PR fixes custom op bug of double backward, follows the same idea of pre-PR.

**Other change:** 
- Fix test case of `double_grad_dynamic`, original test case cannot test eager mode.
- Fix bug: custom cuda double_grad kernel bug

本 PR 修复了自定义算子二阶反向的 bug，遵循 [[Custom operator] Fix custom operator backward=None bug #48656](https://github.com/PaddlePaddle/Paddle/pull/48656) 相同的思路（前置 PR 修复了一阶反向的 bug）

**其他改动**：
- 修复了 `double_grad_dynamic` 测试用例，原来的单测无法访问到新动态图下的二阶反向 op
- 修复了自定义 cuda 二阶反向 kernel 的 bug

### CODE REVIEW

File: python/paddle/fluid/tests/custom_op/test_custom_relu_op_setup.py [hunk 1]
```diff
@@ -153,7 +153,8 @@ def custom_relu_double_grad_dynamic(func, device, dtype, np_x, use_func=True):
     dx = paddle.grad(
         outputs=[out], inputs=[t], create_graph=True, retain_graph=True
     )
-
+    if in_dygraph_mode():
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/custom_op/test_custom_relu_op_setup.py]: Maybe fix this test, since this test doesn't test double grad in the right way

**Review by JiabinYang — COMMENTED** [review]
one comment

File: python/paddle/fluid/tests/custom_op/test_custom_relu_op_setup.py (hunk 1 above)
Comment (jiahy0825) [reply python/paddle/fluid/tests/custom_op/test_custom_relu_op_setup.py]: Already fix this, please review it again, thanks~

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request merged**